lib/awful/key.lua

```diff
@@ -27,23 +24,54 @@ local key = { mt = {}, hotkeys = {} }
 key.ignore_modifiers = { "Lock", "Mod2" }
 
 --- Convert the modifiers into pc105 key names
-local conversion = {
-    mod4    = "Super_L",
-    control = "Control_L",
-    shift   = "Shift_L",
-    mod1    = "Alt_L",
-}
+local conversion = nil
+
+local function generate_conversion_map()
+    if conversion then return conversion end
+
+    local mods = capi.awesome._modifiers
+    assert(mods)
+
+    conversion = {}
+
+    for mod, keysyms in pairs(mods) do
+        for _, keysym in ipairs(keysyms) do
+            assert(keysym.keysym)
+            conversion[mod] = conversion[mod] or keysym.keysym
+            conversion[keysym.keysym] = mod
+        end
+    end
+
+    return conversion
+end
+
+capi.awesome.connect_signal("xkb::group_changed", function() conversion = nil end)
+capi.awesome.connect_signal("xkb::map_changed"  , function() conversion = nil end)
 
 --- Execute a key combination.
 -- If an awesome keybinding is assigned to the combination, it should be
 -- executed.
+--
+-- To limit the chances of accidentally leaving a modifier key locked when
+-- calling this function from a keybinding, make sure is attached to the
+-- release event and not the press event.
+--
 -- @see root.fake_input
 -- @tparam table mod A modified table. Valid modifiers are: Any, Mod1,
 --   Mod2, Mod3, Mod4, Mod5, Shift, Lock and Control.
 -- @tparam string k The key
 function key.execute(mod, k)
+    local modmap = generate_conversion_map()
+    local active = capi.awesome._active_modifiers
+
+    -- Release all modifiers
+    for _, m in ipairs(active) do
+        assert(modmap[m])
+        root.fake_input("key_release", modmap[m])
```

Comment:
This so does not work. If I press right shift and you fake a release of left shift, then right shift is still pressed.

Reply:
Also, I wonder if this even works... I know that xinput-wise, the xtest extension uses its own keyboard. Thus, there would be one input device where left shift is pressed and another input device where a key that was never pressed is released. If I were to guess, I'd say that this means that the fake release just does not do anything, but that's just a guess.

Reply:
It depends on how the modifiers were pressed. For recursive `.execute` and other few case: "it works more than it did before". It isn't the kind of answer you like, but the code is there and it is technically an improvement. Now it works 95.5% of the time instead of 95% of the time.